modules/phase_field/src/auxkernels/PFFractureBoundsAux.C

```diff
@@ -0,0 +1,55 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PFFractureBoundsAux.h"
+#include "SystemBase.h"
+
+registerMooseObject("PhaseFieldApp", PFFractureBoundsAux);
+
+template <>
+InputParameters
+validParams<PFFractureBoundsAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addClassDescription("Provides the upper and lower bound of the phase field fracture "
+                             "variable to PETSc's SNES variational inequalities solver.");
+  params.addRequiredParam<Real>("upper", "The upper bound for the variable");
+  params.addRequiredParam<NonlinearVariableName>("bounded_variable", "The variable to be bounded");
+  return params;
+}
+
+PFFractureBoundsAux::PFFractureBoundsAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _upper_vector(_nl_sys.getVector("upper_bound")),
+    _lower_vector(_nl_sys.getVector("lower_bound")),
+    _bounded_var(_nl_sys.getVariable(_tid, getParam<NonlinearVariableName>("bounded_variable"))),
+    _bounded_var_name(parameters.get<NonlinearVariableName>("bounded_variable")),
+    _upper(getParam<Real>("upper"))
+{
+  if (!isNodal())
```

Comment:
You can something like this here:

```cpp
  const std::vector<std::string> & solver_options =
      _app.getExecutioner()->getParamTempl<std::vector<std::string>>("petsc_options_value");
  if (std::find(solver_options.begin(), solver_options.end(), "vinewtonrsls") ==
          solver_options.end() &&
      std::find(solver_options.begin(), solver_options.end(), "vinewtonssls") ==
          solver_options.end())
    mooseError("A variational solver, i.e. vinewtonrsls, must be used in conjunction with PFFractureBoundsAux");
```

Reply:
But this should be in the base class if you decide to derive it from BoundsAux. This also doesn't work with command line petsc options I guess.

Reply:
The actual right API is `petscOptionsDatabase`() in FEProblemBase, but it requires PETSc-3.12.x or higher.

It is better to add this check in the base class.

Reply:
How to make it backwards-compatible if PETSc version is lower than 3.12.x?

Reply:
The API `petscOptionsDatabase` includes options from both the input file and the command line. It requires 3.12.x or higher.

For old PETSc, you need to use the old way like:

(1) `PetscOptions & petsc = problem.getPetscOptions()` for the input file options

(2) `PetscGetArgs(&argc, &args);` for the comandline options.

PETSc-3.12.x will be the default solver within one month in MOOSE.

Reply:
Great.  Thanks!

Reply:
It may be easier and also better to add a routine `isSNESVI`() in `PetscSupport.C` . `PetscSupport.C` has some reference routines you could look at.